File: core/compressor/CompressionTask.cpp
#include "core/compressor/CompressionTask.h"

#include "system/filesystem/Filesystem.h"

#include <chrono>
#include <ctime>
#include <iomanip>
#include <sstream>

namespace gsm::core {
namespace {

std::string utcNow()
{
    const auto now = std::chrono::system_clock::now();
    const auto time = std::chrono::system_clock::to_time_t(now);
    std::ostringstream stream;
    stream << std::put_time(std::gmtime(&time), "%Y-%m-%dT%H:%M:%SZ");
    return stream.str();
}

} // namespace

CompressionResult Compressor::compress(
    const GameAnalysis& analysis,
    const CompressionRecommendation& recommendation,
    SafetyMetadataStore& metadataStore,
    std::function<void(size_t)> onProgress) const
{
    CompressionResult result;
    result.bytesBefore = analysis.totalBytes;

    if (!analysis.isValid) {
        result.errorMessage = "Cannot compress: analysis is invalid.";
        return result;
    }

    if (recommendation.action != RecommendationAction::Compress) {
        result.errorMessage = "Cannot compress: game is marked as Skip.";
        return result;
    }

    if (!recommendation.algorithm.has_value()) {
        result.errorMessage = "Cannot compress: no algorithm selected.";
        return result;
    }

    if (!gsm::system::directoryExists(analysis.rootPath)) {
        result.errorMessage = "Target path no longer exists.";
        return result;
    }

    const auto algorithm = *recommendation.algorithm;

    SafetyMetadata metadata = metadataStore.createPlannedMetadata(
        analysis,
        recommendation,
        analysis.gameName,
        "manual"
    );
    metadata.state = SafetyOperationState::Running;
    metadata.createdAtUtc = utcNow();
    metadata.updatedAtUtc = metadata.createdAtUtc;

    if (!metadataStore.save(metadata)) {
        result.errorMessage = "Failed to save safety metadata before compression.";
        return result;
    }

    size_t lineCount = 0;
    auto outputCallback = [&lineCount, onProgress](const std::string& chunk) {
        if (!onProgress) return;
        for (char c : chunk) {
            if (c == '\n') {
                ++lineCount;
            }
        }
        onProgress(lineCount);
    };

    const auto command = adapter_.buildCompressCommand(analysis.rootPath, algorithm);
    const auto processResult = adapter_.run(command, outputCallback);

    result.exitCode = processResult.exitCode;
    result.output = processResult.output;
    result.metrics = gsm::system::CompactProcessAdapter::parseCompressOutput(processResult.output);

    if (processResult.exitCode != 0) {
        metadata.state = SafetyOperationState::Failed;
        metadata.notes.push_back("compact.exe exit code: " + std::to_string(processResult.exitCode));
        metadata.updatedAtUtc = utcNow();
        metadataStore.save(metadata);
        result.errorMessage = "compact.exe returned non-zero exit code.";
        return result;
    }

    const auto& metrics = result.metrics;
    if (metrics.parsed && analysis.totalBytes > 0) {
        const std::uintmax_t unchangedBytes = analysis.totalBytes - metrics.bytesBeforeFromOutput;
        metadata.sizeAfterBytes = unchangedBytes + metrics.bytesAfterFromOutput;
    } else {
        metadata.sizeAfterBytes = gsm::system::directorySize(analysis.rootPath);
    }
    metadata.state = SafetyOperationState::Completed;
    metadata.updatedAtUtc = utcNow();

    if (!metadataStore.save(metadata)) {
        result.errorMessage = "Compression succeeded but failed to save metadata.";
        result.success = true;
        result.bytesAfter = metadata.sizeAfterBytes;
        return result;
    }

    result.success = true;
    result.bytesAfter = metadata.sizeAfterBytes;
    return result;
}

CompressionResult Compressor::restore(
    const SafetyMetadata& metadata,
    SafetyMetadataStore& metadataStore,
    const gsm::system::Path& targetPath,
    std::function<void(size_t)> onProgress) const
{
    CompressionResult result;
    result.bytesBefore = metadata.sizeAfterBytes;

    if (!gsm::system::directoryExists(targetPath)) {
        result.errorMessage = "Target path no longer exists.";
        return result;
    }

    size_t lineCount = 0;
    auto outputCallback = [&lineCount, onProgress](const std::string& chunk) {
        if (!onProgress) return;
        for (char c : chunk) {
            if (c == '\n') {
                ++lineCount;
            }
        }
        onProgress(lineCount);
    };

    const auto command = adapter_.buildRestoreCommand(targetPath);
    const auto processResult = adapter_.run(command, outputCallback);

    result.exitCode = processResult.exitCode;
    result.output = processResult.output;

    if (processResult.exitCode != 0) {
        SafetyMetadata updated = metadata;
        updated.state = SafetyOperationState::Failed;
        updated.notes.push_back("Restore failed. compact.exe exit code: " + std::to_string(processResult.exitCode));
        metadataStore.save(updated);
        result.errorMessage = "compact.exe returned non-zero exit code during restore.";
        return result;
    }

    result.bytesAfter = gsm::system::directorySize(targetPath);
    result.success = true;

    SafetyMetadata updated = metadata;
    updated.state = SafetyOperationState::Restored;
    updated.sizeAfterBytes = result.bytesAfter;
    updated.notes.push_back("Restored to original state.");
    metadataStore.save(updated);

    return result;
}

} // namespace gsm::core

File: game_storage_manager/core/compressor/compression_task.py
from __future__ import annotations

import threading
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Callable

from game_storage_manager.core.analyzer.game_analysis import GameAnalysis
from game_storage_manager.core.rules_engine.recommendation_engine import (
    CompressionRecommendation,
    RecommendationAction,
)
from game_storage_manager.core.safety.safety_metadata import SafetyMetadata, SafetyOperationState
from game_storage_manager.core.safety.safety_metadata_store import SafetyMetadataStore
from game_storage_manager.system.filesystem.filesystem import directory_exists, directory_size
from game_storage_manager.system.process.compact_process_adapter import (
    CompactOutputMetrics,
    CompactProcessAdapter,
)


@dataclass
class CompressionResult:
    success: bool = False
    error_message: str = ""
    bytes_before: int = 0
    bytes_after: int = 0
    exit_code: int = -1
    output: str = ""
    metrics: CompactOutputMetrics | None = None


def _utc_now() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


class Compressor:
    def __init__(self):
        self._adapter = CompactProcessAdapter()

    def compress(
        self,
        analysis: GameAnalysis,
        recommendation: CompressionRecommendation,
        metadata_store: SafetyMetadataStore,
        on_progress: Callable[[int], None] | None = None,
        cancel_flag: threading.Event | None = None,
    ) -> CompressionResult:
        result = CompressionResult()
        result.bytes_before = analysis.total_bytes

        if not analysis.is_valid:
            result.error_message = "Cannot compress: analysis is invalid."
            return result

        if recommendation.action != RecommendationAction.Compress:
            result.error_message = "Cannot compress: game is marked as Skip."
            return result

        if recommendation.algorithm is None:
            result.error_message = "Cannot compress: no algorithm selected."
            return result

        if not directory_exists(analysis.root_path):
            result.error_message = "Target path no longer exists."
            return result

        metadata = metadata_store.create_planned_metadata(
            analysis, recommendation, analysis.game_name, "manual"
        )
        metadata.state = SafetyOperationState.Running
        metadata.created_at_utc = _utc_now()
        metadata.updated_at_utc = metadata.created_at_utc

        if not metadata_store.save(metadata):
            result.error_message = "Failed to save safety metadata before compression."
            return result

        line_count = 0

        def output_callback(chunk: str) -> None:
            nonlocal line_count
            if not on_progress:
                return
            line_count += chunk.count("\n")
            on_progress(line_count)

        command = self._adapter.build_compress_command(analysis.root_path, recommendation.algorithm)
        process_result = self._adapter.run(command, output_callback, cancel_flag)

        result.exit_code = process_result.exit_code
        result.output = process_result.output
        result.metrics = CompactProcessAdapter.parse_compress_output(process_result.output)

        if process_result.exit_code != 0:
            metadata.state = SafetyOperationState.Failed
            metadata.notes.append(f"compact.exe exit code: {process_result.exit_code}")
            metadata.updated_at_utc = _utc_now()
            metadata_store.save(metadata)
            result.error_message = "compact.exe returned non-zero exit code."
            return result

        if result.metrics and result.metrics.parsed and analysis.total_bytes > 0:
            unchanged_bytes = analysis.total_bytes - result.metrics.bytes_before_from_output
            metadata.size_after_bytes = unchanged_bytes + result.metrics.bytes_after_from_output
        else:
            metadata.size_after_bytes = directory_size(analysis.root_path)

        metadata.state = SafetyOperationState.Completed
        metadata.updated_at_utc = _utc_now()

        if not metadata_store.save(metadata):
            result.error_message = "Compression succeeded but failed to save metadata."
            result.success = True
            result.bytes_after = metadata.size_after_bytes
            return result

        result.success = True
        result.bytes_after = metadata.size_after_bytes
        return result

    def restore(
        self,
        metadata: SafetyMetadata,
        metadata_store: SafetyMetadataStore,
        target_path: str,
        on_progress: Callable[[int], None] | None = None,
        cancel_flag: threading.Event | None = None,
    ) -> CompressionResult:
        result = CompressionResult()
        result.bytes_before = metadata.size_after_bytes

        if not directory_exists(target_path):
            result.error_message = "Target path no longer exists."
            return result

        line_count = 0

        def output_callback(chunk: str) -> None:
            nonlocal line_count
            if not on_progress:
                return
            line_count += chunk.count("\n")
            on_progress(line_count)

        command = self._adapter.build_restore_command(target_path)
        process_result = self._adapter.run(command, output_callback, cancel_flag)

        result.exit_code = process_result.exit_code
        result.output = process_result.output

        if process_result.exit_code != 0:
            updated = SafetyMetadata(
                schema_version=metadata.schema_version,
                id=metadata.id,
                game_name=metadata.game_name,
                root_path=metadata.root_path,
                source=metadata.source,
                size_before_bytes=metadata.size_before_bytes,
                size_after_bytes=metadata.size_after_bytes,
                file_count_before=metadata.file_count_before,
                algorithm=metadata.algorithm,
                state=SafetyOperationState.Failed,
                created_at_utc=metadata.created_at_utc,
                updated_at_utc=metadata.updated_at_utc,
                notes=list(metadata.notes),
            )
            updated.notes.append(f"Restore failed. compact.exe exit code: {process_result.exit_code}")
            metadata_store.save(updated)
            result.error_message = "compact.exe returned non-zero exit code during restore."
            return result

        result.bytes_after = directory_size(target_path)
        result.success = True

        updated = SafetyMetadata(
            schema_version=metadata.schema_version,
            id=metadata.id,
            game_name=metadata.game_name,
            root_path=metadata.root_path,
            source=metadata.source,
            size_before_bytes=metadata.size_before_bytes,
            size_after_bytes=result.bytes_after,
            file_count_before=metadata.file_count_before,
            algorithm=metadata.algorithm,
            state=SafetyOperationState.Restored,
            created_at_utc=metadata.created_at_utc,
            updated_at_utc=metadata.updated_at_utc,
            notes=list(metadata.notes),
        )
        updated.notes.append("Restored to original state.")
        metadata_store.save(updated)

        return result
